emukit/core/optimization/local_search_acquisition_optimizer.py

```diff
@@ -0,0 +1,110 @@
+import logging
+from typing import List, Tuple
+
+import numpy as np
+
+from emukit.core import (CategoricalParameter, OneHotEncoding, OrdinalEncoding,
+                         ParameterSpace)
+from emukit.core.acquisition import Acquisition
+from emukit.core.optimization.acquisition_optimizer import \
+    AcquisitionOptimizerBase
+from emukit.experimental_design.model_free.random_design import RandomDesign
+
+_log = logging.getLogger(__name__)
+
+
+class LocalSearchAcquisitionOptimizer(AcquisitionOptimizerBase):
+    """ Optimizes the acquisition function """
+    def __init__(self, space: ParameterSpace, num_steps: int, num_samples: int, **kwargs) -> None:
+        """
+        :param space: The parameter space spanning the search problem.
+        :param num_steps: Maximum number of steps to follow from each start point.
+        :param num_samples: Number of initial sampled points where the local search starts.
+        """
+        self.space = space
+        self.num_steps = num_steps
+        self.num_samples = num_samples
+        self.sampler = RandomDesign(space)
+
+    def _neighbours_per_parameter(self, all_features: np.ndarray) -> List[np.ndarray]:
+        assert all_features.ndim == 1, "Expected 1d array, got {}d".format(all_features.ndim)
+
+        parameters = self.space.parameters
+        neighbours = []
+        current_feature = 0
+        for parameter in parameters:
+            features = parameter.round(
+                all_features[current_feature:(current_feature + parameter.dimension)]
+                .reshape(1, -1)).ravel()
+            if isinstance(parameter, CategoricalParameter):
+                if isinstance(parameter.encoding, OrdinalEncoding):
+                    left_right = np.unique([parameter.encoding.round_row(features - 1),
+                                            parameter.encoding.round_row(features + 1)])
+                    neighbours.append(left_right[left_right != features].reshape(-1, 1))
+                elif isinstance(parameter.encoding, OneHotEncoding):
+                    neighbours.append(parameter.encodings[
+                        (parameter.encodings != features).any(axis=1)])
+                else:
+                    raise NotImplementedError("Parameter encoding {} not supported."
+                                              .format(type(parameter.encoding)))
+            else:
+                raise NotImplementedError("Parameter type {} not supported."
+                                          .format(type(parameter)))
+            current_feature += parameter.dimension
+        return neighbours
+
+    def _neighbours(self, all_features: np.ndarray) -> np.ndarray:
```

Comment:
Would be great to have even private methods documented. For example I am immediately clear what this one does, and it does a lot